x/blockdb/database.go

```diff
@@ -0,0 +1,1319 @@
+		}
+	}
+
+	header := s.header
+
+	// Update the header with the current state of the database.
+	header.NextWriteOffset = s.nextDataWriteOffset.Load()
+	heights := s.getBlockHeights()
+	header.MaxContiguousHeight = heights.maxContiguousHeight
+	header.MaxHeight = heights.maxBlockHeight
+	headerBytes, err := header.MarshalBinary()
+	if err != nil {
+		return fmt.Errorf("failed to serialize header for writing state: %w", err)
+	}
+	if uint64(len(headerBytes)) != sizeOfIndexFileHeader {
+		return fmt.Errorf("internal error: serialized header state size %d, expected %d", len(headerBytes), sizeOfIndexFileHeader)
+	}
+
+	if _, err := s.indexFile.WriteAt(headerBytes, 0); err != nil {
+		return fmt.Errorf("failed to write header state to index file: %w", err)
+	}
+	return nil
+}
+
+func (s *Database) getBlockHeights() *blockHeights {
+	heights := s.blockHeights.Load()
+	if heights == nil {
+		return &blockHeights{
+			maxBlockHeight:      unsetHeight,
+			maxContiguousHeight: unsetHeight,
+		}
+	}
+	return heights
+}
+
+// recover detects and recovers unindexed blocks by scanning data files and updating the index.
+// It compares the actual data file sizes on disk with the indexed data size to detect
+// blocks that were written but not properly indexed.
+// For each unindexed block found, it validates the block, then
+// writes the corresponding index entry and updates block height tracking.
+func (s *Database) recover() error {
+	dataFiles, maxIndex, err := s.listDataFiles()
+	if err != nil {
+		return fmt.Errorf("failed to list data files for recovery: %w", err)
+	}
+
+	if len(dataFiles) == 0 {
+		return nil
+	}
+
+	if s.header.MaxDataFileSize == math.MaxUint64 && len(dataFiles) > 1 {
+		return fmt.Errorf("%w: only one data file expected when MaxDataFileSize is max uint64, got %d files with max index %d", ErrCorrupted, len(dataFiles), maxIndex)
+	}
+
+	// ensure no data files are missing
+	// If any data files are missing, we would need to recalculate the max height
+	// and max contiguous height. This can be supported in the future but for now
+	// to keep things simple, we will just error if the data files are not as expected.
+	for i := 0; i <= maxIndex; i++ {
+		if _, exists := dataFiles[i]; !exists {
+			return fmt.Errorf("%w: data file at index %d is missing", ErrCorrupted, i)
+		}
+	}
+
+	// Calculate the expected next write offset based on the data on disk.
+	var calculatedNextDataWriteOffset uint64
+	fileSizeContribution, err := safemath.Mul(uint64(maxIndex), s.header.MaxDataFileSize)
+	if err != nil {
+		return fmt.Errorf("calculating file size contribution would overflow: %w", err)
+	}
+	calculatedNextDataWriteOffset = fileSizeContribution
+
+	lastFileInfo, err := os.Stat(dataFiles[maxIndex])
+	if err != nil {
+		return fmt.Errorf("failed to get stats for last data file %s: %w", dataFiles[maxIndex], err)
+	}
+	calculatedNextDataWriteOffset, err = safemath.Add(calculatedNextDataWriteOffset, uint64(lastFileInfo.Size()))
```

Comment:
I don't understand why we're doing it like this instead of just summing all the file sizes. 
If we have a file in the middle that we can't write to its full extent and we open a new file, the file size will be smaller than the maximum file size, and this computation won't be accurate anymore, or am I missing something?